### PR TITLE
Support resize and watermark in ordering.

### PR DESCRIPTION
**Pull request by shtse8**
| Q                  | A
| ------------------ | ---
| Bug fix?           | no
| New feature?       | yes
| BC breaks?         | no
| Deprecations?      | no
| Fixed tickets      | fixes #issuenum
| Related issues/PRs | #308
| License            | MIT

#### What's in this PR?

Support resize and watermark in ordering.

#### Why?

Cannot resize image then adding watermark. The ordering is ignored.

#### Example Usage

The following codes will have different results.

```php
$ffmpeg = FFMpeg\FFMpeg::create();
$video = $ffmpeg->open($filepath);
$filters = $video->filters();
$filters->resize(new FFMpeg\Coordinate\Dimension($width, $height), FFMpeg\Filters\Video\ResizeFilter::RESIZEMODE_SCALE_HEIGHT, false);
$filters->watermark($watermarkPath, array(
	'position' => 'relative',
	'bottom' => 0,
	'left' => 0,
));
$video->save(new FFMpeg\Format\Video\X264(), $videoPath);
```

```php
$ffmpeg = FFMpeg\FFMpeg::create();
$video = $ffmpeg->open($filepath);
$filters = $video->filters();
$filters->watermark($watermarkPath, array(
	'position' => 'relative',
	'bottom' => 0,
	'left' => 0,
));
$filters->resize(new FFMpeg\Coordinate\Dimension($width, $height), FFMpeg\Filters\Video\ResizeFilter::RESIZEMODE_SCALE_HEIGHT, false);
$video->save(new FFMpeg\Format\Video\X264(), $videoPath);
```

#### To Do

- [x] Create tests

### CODE REVIEW

**Comment by shtse8** [comment]
I have tested some cases and they all works like a charm. But I haven't any experience in creating tests in GitHub. Please help me to create some tests.

**Comment by jens1o** [comment]
Anything is alright(but I would remove the 2 comments). It's easy fix these tests:

1. Go to ResizeFilterTest.php
2. Do what you did in your pr:
    - Replace the commands you deleted with what you added

For example:
https://github.com/PHP-FFMpeg/PHP-FFMpeg/blob/master/tests/Unit/Filters/Video/ResizeFilterTest.php#L45 should be  `array(new Dimension(320, 240), ResizeFilter::RESIZEMODE_FIT, 640, 480, 2, array('-vf', '[in]scale=320:240 [out]')),`

**Comment by shtse8** [comment]
Thanks! I have corrected the tests.

**Comment by jens1o** [comment]
Let's look wether we missed something. 😉 This takes about 20 minutes. Don't panic...

**Comment by jens1o** [comment]
Okey, something went wrong with encoding... I don't know why.

**Comment by shtse8** [comment]
It's because some `apply` returns are just using simple syntax such as `transpose=1`. But I had assumed everyone is coding like `[in]transpose=1 [out]`, so i failed to merge each filter into one.
I am trying to fix. Hope this fix can help.

**Comment by jens1o** [comment]
it works! Great! Well done. Now, it is time to wait until @romainneutron had a look.

File: src/FFMpeg/Filters/Video/ResizeFilter.php [hunk 1]
```diff
@@ -98,8 +98,10 @@ public function apply(Video $video, VideoInterface $format)
         if (null !== $dimensions) {
             $dimensions = $this->getComputedDimensions($dimensions, $format->getModulus());
 
-            $commands[] = '-s';
-            $commands[] = $dimensions->getWidth() . 'x' . $dimensions->getHeight();
+            // Using Filter to have ordering
+            $commands[] = '-vf';
+            $commands[] = '[in]scale=' . $dimensions->getWidth() . ':' . $dimensions->getHeight() . ' [out]';
```
Comment (Romain) [inline src/FFMpeg/Filters/Video/ResizeFilter.php]: Do we have to put [int] and [out]?
[This doc](https://trac.ffmpeg.org/wiki/Scaling%20(resizing)%20with%20ffmpeg) don't use them.

File: src/FFMpeg/Media/Video.php [hunk 2]
```diff
@@ -129,6 +129,56 @@ public function save(FormatInterface $format, $outputPathfile)
...
+            }
+            
+            $videoFilterCommands[] = $command;
+        }
+        $videoFilterCommand = implode(";", $videoFilterCommands);
+        
+        if ( $videoFilterCommand ) {
+            $commands[] = '-vf';
+            $commands[] = $videoFilterCommand;
+        }
```
Comment (Romain) [inline src/FFMpeg/Media/Video.php]: Hi @shtse8,

Thanks for this PR.

What are you trying to do here?
Why are you defining a command like this: `[in]scale=...[out]` to get rid of the [in] and [out] afterwards?
I think you should also wrap this in a test `if($commands !== NULL)` and use foreach instead of for.

To my opinion, this part is useless. The command generated by the filter should already be clean enough to be used by the save function, as is, without being processed in another way.

Can you paste here the desired command? Is it somethink like `ffmpeg -i input.avi -vf scale=320:240 output.avi`?
If yes, I think you don't need all this.

Comment (shtse8) [inline src/FFMpeg/Media/Video.php]: because mutiple vf is not allow in one command. if you want to process the video with ordering . you need to merge mutiple vf into one with ordering.

File: src/FFMpeg/Filters/Video/ResizeFilter.php (hunk 1 above)
Comment (shtse8) [inline src/FFMpeg/Filters/Video/ResizeFilter.php]: because it is used to be merged outside . it is the best practice to have in and out in filter command .
but if you have only one filter . you can just ignore the in and out. because it must be in in front and out behind. but we need to do the orderinga and multiple filters. it s better to specify the in and out.

File: src/FFMpeg/Media/Video.php (hunk 2 above)
Comment (shtse8) [inline src/FFMpeg/Media/Video.php]: but merging the filters into one is not an easy job. because the structure of the class is now returning the command directly. that mean mutiple vf command is returned. we have to find a way to merge then instead to reconstruct the whole class which is a huge change of the class. this is the best i think without any problem because all commands are returned by the classes, it is easy for us to know the format and merge them into one.

Comment (shtse8) [inline src/FFMpeg/Media/Video.php]: `foreach` is not better then `for` because we need to take two arguments if matching the `vf` command.

Comment (shtse8) [inline src/FFMpeg/Media/Video.php]: `foreach` is hard to control the index resulting in poor performance.

Comment (shtse8) [inline src/FFMpeg/Media/Video.php]: the commands generated by the filter is not clean enough to be used by the `save` function. we need to do the ordering, `vf` is the procedure to do the ordering filter. you are hard to generate only one perfer command as only one `vf` command is allowed but you have mutiple filter classes generating multiple `vf` command. so now we have no choice, one is to modify all classes to generate a special return for `vf`, and one is to merge multiple `vf` into one in the preprocess. i think the second way is the best for our situation.

Comment (shtse8) [inline src/FFMpeg/Media/Video.php]: the desired command is not just that simple.
because we need to scale first and then adding watermark.
the desired command is 
`ffmpeg -i input.avi -vf "movie=watermark.png [watermark]; [in][watermark] overlay=100:100 [p1]; [p1] scale=320:240 [out]" output.avi`
Also, because the pull request is to support multiple filters with ordering.
the desired command can be more complex as follow:
`ffmpeg -i input.avi -vf "movie=watermark.png [watermark]; [in][watermark] overlay=100:100 [p1]; [p1] scale=320:240 [p2]; [p2] [watermark] overlay=200:200 [p3]; [p3] scale=100:100 [out]" output.avi`

File: src/FFMpeg/Filters/Video/ResizeFilter.php (hunk 1 above)
Comment (shtse8) [inline src/FFMpeg/Filters/Video/ResizeFilter.php]: the doc is just simplify for basic user. but for the concept of `vf` is to control the `in` and `out` and then adding filters step by step.
`-v scale=100:100` means `-v "[in] scale=100:100 [out]"` actually.
if you want to scale two times, you will have a command like this:
`-v "[in] scale=100:100 [p1]; [p1] scale=200:200 [out]"` but you cannot do `-v "scale=100:100; scale=200:200" as you are missing the in and out. the filter processor don't know how to do.

Comment (shtse8) [inline src/FFMpeg/Filters/Video/ResizeFilter.php]: Reference: http://stackoverflow.com/questions/10934420/ffmpeg-how-to-scale-a-video-then-apply-a-watermark

Comment (shtse8) [inline src/FFMpeg/Filters/Video/ResizeFilter.php]: `p1` is a custom variable defined by me.
so we can easy to merge multiple filters with the variables p1, p2, p3, p4 .... p100 somethings like this.

Comment (Romain) [inline src/FFMpeg/Filters/Video/ResizeFilter.php]: Ok, thanks for the explanation.

File: src/FFMpeg/Media/Video.php (hunk 2 above)
Comment (Romain) [inline src/FFMpeg/Media/Video.php]: Ok, thanks for the explanation.

**Comment by Romain** [comment]
Ok, I'm fine with the idea.
Can you just update the README file with the new way to use this class?
Thanks for this

**Comment by alexking** [comment]
There's a gotcha with this, if you're making a Gif, that uses its own `save()` method which does not merge `-vf` calls, and uses one of its own to do scaling, so you basically can't use `-vf` filters with it.
